Add header-sensitive size checking via CPubKey::ValidSize, and prefer key/sig size const references

Empact opened the pull request:
Here I:
* add header-sensitive size checking via CPubKey::ValidSize
* move toward referring to the consts where possible, rather than magic numbers
* rename size consts to not repeat the type in variable name

My first PR! Just getting a sense of the codebase at this point.
The last point at least is purely a question of style, happy to back that out.

Empact committed:
Extract CPubKey::SIGNATURE_SIZE and COMPACT_SIGNATURE_SIZE

As CPubKeySig::SIZE and COMPACT_SIZE

Empact committed:
Rename CPubKey::PUBLIC_KEY_SIZE and COMPRESSED_PUBLIC_KEY_SIZE

As SIZE and COMPRESSED_SIZE

Empact committed:
Assert CPubKey::ValidLength to the pubkey's header-relevent size

Previously this was an inline test where the specificity was probably judged
overly specific. As a class method it makes sense to maintain consistency.

Empact committed:
Rename CKey::PRIVATE_KEY_SIZE and COMPRESSED_PRIVATE_KEY_SIZE

To SIZE and COMPRESSED_SIZE

devrandom reviewed (commented):
Although this is good cleanup overall, I would revert the change to `IsValid`.  See inline comment.

Empact commented on src/pubkey.h:
```diff
@@ -164,7 +172,7 @@ class CPubKey
     bool IsValid() const
     {
-        return size() > 0;
+        return size() >= COMPRESSED_SIZE && size() <= SIZE;
```
Roger! Thanks for pointing that out.

Empact committed:
Assert compressed / compact keys and sigs are strictly shorter than regular

This is currently true and intuitively expected

Reword COMPRESSED_SIZE static assert message to make the desired arrangement
explicit.

Empact commented on src/script/interpreter.cpp:
```diff
@@ -61,17 +61,17 @@ static inline void popstack(std::vector<valtype>& stack)
 }
 
 bool static IsCompressedOrUncompressedPubKey(const valtype &vchPubKey) {
```
Make sense! This commit moves these methods: https://github.com/Empact/bitcoin/commit/bd90b1a5f063a43fbf7509d0fe8760bb38594fc5

I'll plan to submit it once this is merged.

devrandom commented:
CI failed...

Empact commented:
@devrandom Fixed

devrandom commented:
ACK https://github.com/bitcoin/bitcoin/pull/12319/commits/7624bc9dc0c04790448504923ea1827776b701d7

Empact commented:
Split this up for easier review. Closing in favor of #12461 #12460 and #12459.
